pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,371 @@
+  // Once a partition is marked for verification, we wait 10 minutes to pull its ideal state.
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this amount of time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 100;
+
+  /*
+   * Class to keep an ingestion delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase ingestion delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  static private class DelayMeasure {
+    public DelayMeasure(long t, long d) {
+      _delayMilliseconds = d;
+      _sampleTime = t;
+    }
+    public final long _delayMilliseconds;
+    public final long _sampleTime;
+  }
+
+  // HashMap used to store delay measures for all partitions active for the current table.
+  private final ConcurrentHashMap<Integer, DelayMeasure> _partitionToDelaySampleMap = new ConcurrentHashMap<>();
+  // We mark partitions that go from CONSUMING to ONLINE in _partitionsMarkedForVerification: if they do not
+  // go back to CONSUMING in some period of time, we confirm whether they are still hosted in this server by reading
+  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
+  private final ConcurrentHashMap<Integer, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
+
+  final int _timerThreadTickIntervalMs;
+  // Timer task to check partitions that are inactive against ideal state.
+  private final Timer _timer;
+
+  private final ServerMetrics _serverMetrics;
+  private final String _tableNameWithType;
+  private final String _metricName;
+
+  private final boolean _enablePerPartitionMetric;
+  private final boolean _enableAggregateMetric;
+  private final Logger _logger;
+
+  private final RealtimeTableDataManager _realTimeTableDataManager;
+  private Clock _clock;
+
+  /*
+   * Returns the maximum ingestion delay amongs all partitions we are tracking.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (int partitionGroupId : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(partitionGroupId);
+      if ((newMax == null)
+          ||
+          (currentMeasure != null) && (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  /*
+   * Helper function to age a delay measure. Aging means adding the time elapsed since the measure was
+   * taken till the measure is being reported.
+  *
+  * @param currentDelay original sample delay to which we will add the age of the measure.
+   */
+  private long getAgedDelay(DelayMeasure currentDelay) {
```

Comment:
Suggestion:
You can remove this method, in favor of a method inside the DelayMeasure class that computes the agedDelay. You can do the null check outside this method in the one case where it may happen that it is null.

Reply:
This is higher abstraction than plain delay concept, doing the suggested will also make me pass the clock to delay which is currently very simple. I would rather leave as is.